lib/posix/eventfd.c

```diff
@@ -49,53 +62,72 @@ static int eventfd_poll_update(struct eventfd *efd,
 {
 	ARG_UNUSED(efd);
 
-	if (pfd->events & ZSOCK_POLLOUT) {
-		pfd->revents |= ZSOCK_POLLOUT;
-	}
-
 	if (pfd->events & ZSOCK_POLLIN) {
 		if ((*pev)->state != K_POLL_STATE_NOT_READY) {
 			pfd->revents |= ZSOCK_POLLIN;
 		}
 		(*pev)++;
 	}
 
+	if (pfd->events & ZSOCK_POLLOUT) {
+		if ((*pev)->state != K_POLL_STATE_NOT_READY) {
+			pfd->revents |= ZSOCK_POLLOUT;
+		}
+		(*pev)++;
+	}
+
 	return 0;
 }
 
 static ssize_t eventfd_read_op(void *obj, void *buf, size_t sz)
 {
 	struct eventfd *efd = obj;
-	eventfd_t count;
-	int ret;
+	int result = 0;
+	eventfd_t count = 0;
+	struct k_spinlock_key key;
 
 	if (sz < sizeof(eventfd_t)) {
 		errno = EINVAL;
 		return -1;
 	}
 
-	ret = k_sem_take(&efd->read_sem,
-			 (efd->flags & EFD_NONBLOCK) ? K_NO_WAIT : K_FOREVER);
+	for (;;) {
+		key = k_spin_lock(&efd->lock);
+		if ((efd->flags & EFD_NONBLOCK) && efd->cnt == 0) {
+			result = EAGAIN;
+			break;
+		} else if (efd->cnt == 0) {
+			z_pend_curr(&efd->lock, key, &efd->wait_q, K_FOREVER);
+		} else {
+			count = (efd->flags & EFD_SEMAPHORE) ? 1 : efd->cnt;
+			efd->cnt -= count;
+			if (efd->cnt == 0)
+				k_poll_signal_reset(&efd->read_sig);
+			break;
+		}
+	}
+	z_unpend_all(&efd->wait_q);
```

Comment:
probably could be moved to line 106 as it doesn't need to be called in the non-blocking case